carcara/src/ast/term.rs

```diff
@@ -569,6 +576,68 @@ impl From<SortedVar> for Term {
     }
 }
 
+impl Sort {
+    // Whether this sort can be unified with another. The map argument
+    // will be a substitution of sort variables to sorts
+    pub fn match_with(&self, target: &Sort, map: &mut IndexMap<String, Sort>) -> bool {
+        match (self, target) {
+            (Sort::Var(a), _) => {
+                match map.entry(a.to_string()) {
+                    Entry::Vacant(e) => {
+                        e.insert(target.clone());
+                    }
+                    Entry::Occupied(e) => {
+                        return e.get() == target;
+                    }
+                }
+                true
+            }
+            (Sort::Atom(a, sorts_a), Sort::Atom(b, sorts_b)) => {
+                if a != b {
+                    false
+                } else {
+                    let matching = sorts_a
+                        .iter()
+                        .zip(sorts_b.iter())
+                        .filter(|&(t_a, t_b)| {
+                            let s_a = t_a.as_sort().unwrap();
+                            let s_b = t_b.as_sort().unwrap();
+                            s_a.match_with(s_b, map)
+                        })
+                        .count();
+                    matching == sorts_a.len() && matching == sorts_b.len()
+                }
+            }
+            (Sort::Function(sorts_a), Sort::Function(sorts_b)) => {
+                for (a_t, b_t) in sorts_a.iter().zip(sorts_b.iter()) {
+                    let a_s = a_t.as_sort().unwrap();
+                    let b_s = b_t.as_sort().unwrap();
+                    if !a_s.match_with(b_s, map) {
+                        return false;
+                    }
+                }
+                true
```

Comment:
I think you can also use `Iterator::all` here.
```suggestion
                sorts_a.iter().zip(sorts_b.iter()).all(|(a_t, b_t)| {
                    let a_s = a_t.as_sort().unwrap();
                    let b_s = b_t.as_sort().unwrap();
                    a_s.match_with(b_s, map)
                })
```